### PR TITLE
Skins :: fix qss icons with kIconThemes 5.80

### PR DESCRIPTION
**Pull request by ronso0**
# 1

missing skin icons due to broken kIconThemes https://www.ubuntuupdates.org/pm/kiconthemes were fixed
BUT that lib forces all `image: ...` icons to 1:1 aspect ratio while especially LateNight uses different formats.

revival of #3863 

For those affected by https://bugs.launchpad.net/mixxx/+bug/1922966
(Arch, Ubuntu 21.04 with kiconthemes installed)

the `background` shorthand is the only one-line way to achieve the previous look
(alternative: separate `background-image`, `background-position` and `background-repeat` lines...)
``` diff
 #Hotcue1 WPushButton[displayValue="1"][dark="false"] {
-  image: url(skin:/palemoon/buttons/btn__1_active.svg) no-repeat center center;
+  background: url(skin:/palemoon/buttons/btn__1_active.svg) no-repeat center center;
 }
```
BUT since we omit `background-color` transparent bg color is assumed, thus all button background colors must be moved after setting the icons.

In ALL skin
* all icons show up with appropriate size
* all button states have the correct background color
* the logo in the launch screen is sized correctly

_____

# 2

https://github.com/mixxxdj/mixxx/pull/3902/commits/76db8ba6a878398bc25cd00ee0fd7ff853e09c71
(included here because don't want to spend the time to rebase. this PR will be finished soon anyway)
In Tango the template and icon paths are `skin:../Tango/buttons/...` to allow reusing them for Tango64.
After #3877 those paths need to be `skin:/../Tango/buttons/...` so they're correctly resolved to `skins/Tango/buttons/...`

___

# 3

https://github.com/mixxxdj/mixxx/pull/3902/commits/24c3eaf6892e2e7e7dd6ee6c4285e366eea55457
use 'background-image' for the default launch image style

### CODE REVIEW

**Comment by ronso0** [comment]
### ToDo
- [ ] double check Shade
- [ ] fix LateNight **Classic**
- [ ] check Deere (little affected at most)

**Comment by ronso0** [comment]
done.

**Comment by Holzhaus** [comment]
Thanks. I can test this tomorrow, but who can verify that the fix works? That requires testing by someone who can reproduce the issue in the first place. @ywwg? @daschuer?

**Comment by ronso0** [comment]
yes @ywwg in case he didn't patch the icon lib, yet.

I tested this myself with Ubuntu 21.04 and that kIconlib installed. This is mostly about making the LateNight main interface look good and Tango usable again.
there are a few minor glitches left with some expand/collapse icons in Deere for example, or some library preview buttons,..
but the effort fixing those is not worth it -- considering that only a few distros are affected and that lib will hopefully be fixed in a few weeks.

**Comment by ywwg** [comment]
The only issue I see is no bpm lock icon or library triangles on deere 64.  Otherwise good!

**Comment by daschuer** [comment]
I can confirm that the launch image is no longer squeezed. 
The search bar icon is also fixed.
Cue/Play/pause fixed.

The preview button is still squeezed. 
The effect button is still broken. 

How about creating a PR with only the known working fixes? 

I have just contacted the package maintainers Rick and  Jose again. Maybe we can solve the root cause with them. 
And live with the remaining issues until than.

**Comment by ronso0** [comment]
Deere64 icons are now visible with the path fix used for Tango already `url(skin:/icon/any.svg` > `url(skin:/../Deere/icon/any.svg`

ready!

> I have just contacted the package maintainers Rick and Jose again. Maybe we can solve the root cause with them.

They simply need to update the package in the ubuntu main repo to 82. It's already in kubuntu-backport
https://www.ubuntuupdates.org/pm/kiconthemes

> And live with the remaining issues until than.

Only issue I'm aware of are squeezed icons / enforced to 1:1 and not scaled: `<`, `>`, table sort arrows and the previewbutton in the library. Yes, we can live with that..

**Comment by daschuer** [comment]
Can you revert the non working attempt before merge?

**Comment by ronso0** [comment]
sure, I can squash commits per skin tomorrow.

**Comment by ronso0** [comment]
it's not non-working, it's two separate steps btw

**Comment by daschuer** [comment]
There is no need to squash anything. I just want make sure not to have changes that do not improve the situation but have an regression risk.  
During my test I cannot confirm that the preview column is fixed, but there are related changes. The same is true for the effect buttons.

**Comment by ronso0** [comment]
I checked again: the preview and bpm icons are there in all skins.
we didn't you see it?

**Comment by daschuer** [comment]
The preview icon is there but squashed the same goes for the effect combo box.

I have a plain Ubuntu Hirsute with the faulty library installed. I can't reproduce the icon missing issue, only the squashing.

**Review by ywwg — APPROVED** [review]
looks good!

**Comment by ronso0** [comment]
> I can't reproduce the icon missing issue, only the squashing.

that was a regression caused by #3877 affecting Deere64 and Tango64 where (incorrect/incomplete) icon urls in the base skin were not properly resolved. see PR description

> The preview icon is there but squashed the same goes for the effect combo box.

To re-cap the consequneces of that broken lib:
**previously**, we used mostly `image` for all skin icons
* all SVG icons would be displayed and scaled properly, regardless if they're oversized (icon_4px.svg) or not
* the icon size would be taken from the actual size of the SVG, the resizable Preview button the icon (~ 2:1 aspect ratio) would be scaled depending on the column width

**now**, (with broken kIconThemes)
* relative paths for SVG icons were ...ignored? thus provide absolute paths with #3877
* `image` always assume an aspect ratio of 1:1 (distort all non-square icons) thus use `background` now, where we accept the consequences:
* `background: color icon repeat position-x position-y` does actually scale icons to fit Mixxx widgets -- but that doesn't work for Qt sub-controls, like `WEffectSelecotor::down-arrow` or `QHeaderView::up-arrow`, there they're used at real size.
That does matter Deere where almost all icons are 48x48 pixles. So drop-down arrows, sort arrows etc. would be huge when using `background` there. with `image` they're only sueezed.

For the **preview button** I now decided to go with the fixed-size icons with correct aspect ratio set in `default.qss` (included at compile time, icons in mixxx.qrc) which will not grow when the library font is increased
instead of a auto-scaling icon which is squeezed (square).

I don't want to spent hours with resizing icons for a skin I don't care about tbh
So, yes, as you said: there are issues users with libKIcon..80 need to live with until that lib is finally updated

Re squeezed **effect combobox**: please post an example screen.

**Comment by daschuer** [comment]
I like to merge this asap. It is kind of unfair to put so much work on you (us), just because the package maintainers are not able to upload the fixed version to the Hirsute Repro. 

My original concern was not to commit half baked fixes that will hit us later once the library is updated. 

If you can confirm that this is not the case let's just merge this and carry on. Do you?

File: res/skins/Deere/style.qss
Comment (daschuer) [inline res/skins/Deere/style.qss:319]: Does this case a regression for all?
If so I prefere to revert this. The same applies to the other todos.

**Comment by daschuer** [comment]
I have now copied the 5.82 version to our PPA.
Issue fixed :-/

Which changes from this PR are still relevant?

File: res/skins/Deere/style.qss
Comment (ronso0) [reply res/skins/Deere/style.qss:319]: this is the only way to get the Preview icons undistorted and adpating to the column width (though not growing with the row height). It's the same for all skins now, and special button styles are not re-implemented, thus the TODOS

so Yes, a regression. with the fixed lib `image` should work as it did before, and I will revert that asap after the patched libs are in main repos.

**Comment by ronso0** [comment]
> My original concern was not to commit half baked fixes that will hit us later once the library is updated.

The fixed lib should restore the previous working status of `image`, not screw it up even more.
There are no halfbaked fixes, and I don't expect any regressions -- unless the fixed lib scres up `background`, too. but the patch is only about SVG and relative paths, so if I undertstood it correctly the fixed/restired lib shouldn't interfere with our skin stylesheets anymore _at all_. Basically we could revert the entire PR, but 99% of the changes make skins future-proof.

There are a few places (icons in Deere only IIRC, as well as the Preview button) where we should revert the fixes afterwards to make it look good again.

Ready.

**Comment by ronso0** [comment]
> I have now copied the 5.82 version to our PPA.
> Issue fixed :-/

This is not a fix -- it's just a mirror of the kubuntu ppa.
To make use of that/or our ppa users need to
* complain about it somewhere (best case Mixxx forums)
* someone knowing about the reason AND the fix has to be around, and then I'd rather recommend them adding the kubuntu ppa instead

As explained this makes skins more immune against future broken libs, and we should revert the few changes marked with TODO soon. I have this package on the radar and will do so asap.

**Comment by daschuer** [comment]
.

ronso0 ***@***.***> schrieb am Fr., 28. Mai 2021, 16:13:

**Comment by ronso0** [comment]
@daschuer 

From your last response only
> .

is visible.

Ready for merge?

**Comment by daschuer** [comment]
The PPA solution is a automatic fix for all PPA users. 
However, it did not work well, because of missing dependencies of the new package. 
I think we need a patched 5.80 version, instead of a full 5.82 version that depends on others. 
I have deleted that package form our PPA.

**Comment by daschuer** [comment]
So let's merge this now. Thank you for developing the fix.

**Comment by uklotzde** [comment]
Some merge conflicts with main need to be resolved.

**Comment by ronso0** [comment]
I'll take care of that tomorrow.

**Comment by ronso0** [comment]
Done.

**Comment by coolerminds** [comment]
awesome!

**Pull request merged**
